<html><head><title>HW2</title></head><body style="max-width:468pt;background-color:#ffffff;padding:72pt 72pt 72pt 72pt"><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span>Bootcamp HW 2</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="background-color:#ff9900;font-weight:bold">I already download one RNA-seq SRA of your choice with single-end reads:</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="background-color:#ff9900">Navigated to </span><span style="color:#1155cc;background-color:#ff9900;text-decoration:underline"><a href="http://www.google.com/url?q=http%3A%2F%2Fwww.nature.com%2Fencode%2Fcategory%2Fresearch-papers&amp;sa=D&amp;sntz=1&amp;usg=AFQjCNE-C08zjMvRTmj2uYjNoOWVCI2MxA" style="color:inherit;text-decoration:inherit">this tab</a></span><span style="background-color:#ff9900">&nbsp;on the ENCODE nature page, found the five accession numbers from </span><span style="color:#1155cc;background-color:#ff9900;text-decoration:underline"><a href="http://www.google.com/url?q=http%3A%2F%2Fwww.nature.com%2Fnature%2Fjournal%2Fv489%2Fn7414%2Ffull%2Fnature11233.html&amp;sa=D&amp;sntz=1&amp;usg=AFQjCNGrSV49OepnV2VS8ANIFTu67uZECg" style="color:inherit;text-decoration:inherit">this paper</a></span><span style="background-color:#ff9900">, downloaded [files] from </span><span style="color:#1155cc;background-color:#ff9900;text-decoration:underline"><a href="http://www.google.com/url?q=http%3A%2F%2Fwww.ncbi.nlm.nih.gov%2Fgeo%2Fquery%2Facc.cgi%3Facc%3DGSM958739&amp;sa=D&amp;sntz=1&amp;usg=AFQjCNF6yhxJwUT994gC0XZKVXmbx2mq2g" style="color:inherit;text-decoration:inherit">this GEO entry</a></span><span style="background-color:#ff9900">. &nbsp; &nbsp;</span></p><p style="margin-right:0;text-indent:36pt;text-align:justify;color:#000000;direction:ltr;font-size:11pt;margin-left:36pt;margin-bottom:0;font-family:&quot;Arial&quot;;margin-top:0;padding:0"><span style="background-color:#ff9900">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="background-color:#ff9900;font-weight:bold">Converted to fastq</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="background-color:#ff9900">/home/jdavistu/sratoolkit.2.1.16-centos_linux64/fastq-dump </span><span style="font-style:italic;background-color:#ff9900">file.sra </span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="background-color:#ff9900"></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="background-color:#ff9900;font-weight:bold">Built bowtie index</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="background-color:#ff9900">Downloaded hg19 refseq genes: fasta, from UCSC genome browser.</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="background-color:#ff9900">bowtie-build hg19_refseq.fasta hg19_transcriptome </span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold">Bootcamp Homework 2:</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold"></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold">Goal: Align and analyze RNA-seq data</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold"></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold">1) </span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt"></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt">&nbsp; Download some data from ENCODE, </span><span style="font-size:12pt;font-weight:bold">align to transcriptome.</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt">Take reads that </span><span style="font-size:12pt;font-weight:bold">didn&#39;t align, </span><span style="font-size:12pt">align to genome.</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Report these numbers. &nbsp;</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt;text-decoration:underline">Re-use your code from your homework from yesterday (you will have to modify it):</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt;text-decoration:underline"></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt">Make a histogram. &nbsp;How is it different than the one from yesterday?</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt">What would </span><span style="font-style:italic;font-size:12pt;font-weight:bold">you</span><span style="font-style:italic;font-size:12pt">&nbsp;</span><span style="font-size:12pt">consider a good cutoff for calling a gene present? &nbsp;Why?</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt;font-weight:bold">Normalization: read in gene length, compute RPKM</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-size:12pt">Generate a list of genes that are differentially expressed at 24 hours. &nbsp;</span></p></body></html>