<html><head><title>Correct RNAseq procedures</title></head><body style="max-width:468pt;background-color:#ffffff;padding:72pt 72pt 72pt 72pt"><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span>Try running bowtie like this (making bowtie output, not .sam output): </span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold">Remember to run with the options that you would like</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="color:#666666;font-style:italic">/opt/biotools/bowtie/bin/bowtie /home/jdavistu/indices/hg19_transcriptome /home/jdavistu/HW2_data/Rep1.fastq &nbsp;&gt; Rep1.bt</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="color:#666666;font-style:italic"></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="color:#666666;font-style:italic">/opt/biotools/bowtie/bin/bowtie /home/jdavistu/indices/hg19_transcriptome /home/jdavistu/HW2_data/Rep2.fastq &nbsp;&gt; Rep2.bt</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold"></span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold"></span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold"></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold">PYTHON PSEUDO-CODE:</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span># 1. Read in bowtie file </span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span># &nbsp;count how many reads map to each gene,</span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span># &nbsp;output a CSV file that </span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span style="font-weight:bold"></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span>see</span><span style="font-weight:bold">&nbsp;BowtieReader.py</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span># python BowtieReader.py &nbsp;&lsquo;&rsquo;&rsquo;currently it inputs Rep1.bt, and creates Rep1.hits&rsquo;&rsquo;&rsquo;</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p><p style="color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span># 2. Jeremy made a file </span><span style="font-weight:bold">transcript_lengths.txt </span><span>(using this script: python getLens.py &gt; transcript_lengths.txt)</span></p><p style="height:11pt;color:#000000;direction:ltr;font-size:11pt;margin:0;font-family:&quot;Arial&quot;;padding:0"><span></span></p></body></html>